easybuild/easyblocks/j/java.py

```diff
@@ -76,15 +78,128 @@ def extract_step(self):
             adjust_permissions(self.builddir, stat.S_IWUSR, add=True, recursive=True)
 
     def install_step(self):
+        """Custom install step: just copy unpacked installation files."""
         if LooseVersion(self.version) < LooseVersion('1.7'):
             remove_dir(self.installdir)
             copy_dir(os.path.join(self.builddir, 'jdk%s' % self.version), self.installdir)
         else:
             PackedBinary.install_step(self)
 
+    def post_install_step(self):
+        """
+        Custom post-installation step:
+        - ensure correct glibc is used when installing into custom sysroot and using RPATH
+        """
+        super(EB_Java, self).post_install_step()
+
+        # patch binaries and libraries when using alternate sysroot in combination with RPATH
+        sysroot = build_option('sysroot')
+        if sysroot and self.toolchain.use_rpath:
+            if not which('patchelf'):
+                error_msg = "patchelf not found via $PATH, required to patch RPATH section in binaries/libraries"
+                raise EasyBuildError(error_msg)
+
+            try:
+                # list of paths in sysroot to consider for adding to RPATH section
+                sysroot_lib_paths = glob.glob(os.path.join(sysroot, 'lib*'))
+                sysroot_lib_paths += glob.glob(os.path.join(sysroot, 'usr', 'lib*'))
+                sysroot_lib_paths += glob.glob(os.path.join(sysroot, 'usr', 'lib*', 'gcc', '*', '*'))
+                if sysroot_lib_paths:
+                    self.log.info("List of library paths in %s to add to RPATH section: %s", sysroot, sysroot_lib_paths)
+
+                # find path to ELF interpreter
+                elf_interp = None
+
+                for ld_glob_pattern in (r'ld-linux-*.so.*', r'ld*.so.*'):
+                    res = glob.glob(os.path.join(sysroot, 'lib*', ld_glob_pattern))
+                    self.log.debug("Paths for ELF interpreter via '%s' pattern: %s", ld_glob_pattern, res)
+                    if res:
+                        # if there are multiple hits, make sure they resolve to the same paths,
+                        # but keep using the symbolic link, not the resolved path!
+                        real_paths = nub([os.path.realpath(x) for x in res])
+                        if len(real_paths) == 1:
+                            elf_interp = res[0]
+                            self.log.info("ELF interpreter found at %s", elf_interp)
+                            break
+                        else:
+                            raise EasyBuildError("Multiple different unique ELF interpreters found: %s", real_paths)
+
+                if elf_interp is None:
+                    raise EasyBuildError("Failed to isolate ELF interpreter!")
+
+                bindir = os.path.join(self.installdir, 'bin')
```

Comment:
Everything up until now has been independent of the actual package being installed. Rather than an explicit `bin`, can we just check all the dirs that are being added to `PATH`?

Reply:
Are we "guaranteed" to have the additions to `PATH` finalised by this point?

Reply:
Not really, since `make_module_req_guess` and `make_module_extra` are generally responsible for specifying what gets included in `$PATH` (and `$LD_LIBRARY_PATH`).

The way that's done currently is by emitting Tcl or Lua statements to update these environment variables directly, we would need to do some extra bookkeeping to know what's being added to `$PATH` and/or `$LD_LIBRARY_PATH` in a clean way in `post_install_step` already...

Reply:
`make_module_req_guess` just returns a dict, so we can just call it? We also know that `make_module_extra` doesn't handle `PATH` or `LIBRARY_PATH`, unless there is a reason for an exception somewhere

Reply:
(I'm explicitly saying `LIBRARY_PATH` as we filter `LD_LIBRARY_PATH` in EESSI)

Reply:
PR to do this at https://github.com/boegel/easybuild-easyblocks/pull/41